src/Build/Evaluation/Expander.cs

```diff
@@ -3268,6 +3275,130 @@ internal object Execute(object objectInstance, IPropertyProvider<T> properties,
+            /// Binding is expensive and throws first-chance MissingMethodExceptions, which is
+            /// bad for debugging experience and has a performance cost.
+            /// A typical binding operation with exception can take ~1.500 ms; this call is ~0.050 ms
+            /// (rough numbers just for comparison).
+            /// See https://github.com/Microsoft/msbuild/issues/2217
+            /// </summary>
+            /// <param name="objectInstance">Object that the function is called on</param>
+            /// <param name="args">arguments</param>
+            /// <returns>The value returned from the function call</returns>
+            private object ExecuteWellKnownFunction(object objectInstance, object[] args)
+            {
+                if (objectInstance is string)
+                {
+                    string text = (string)objectInstance;
+                    if (string.Equals(_methodMethodName, "Substring", StringComparison.OrdinalIgnoreCase))
+                    {
+                        int startIndex;
+                        int length;
+                        if (TryGetArg(args, out startIndex))
+                        {
+                            return text.Substring(startIndex);
+                        }
+                        else if (TryGetArgs(args, out startIndex, out length))
+                        {
+                            return text.Substring(startIndex, length);
+                        }
+                    }
+                    else if (string.Equals(_methodMethodName, "Split", StringComparison.OrdinalIgnoreCase))
+                    {
+                        string separator;
+                        if (TryGetArg(args, out separator) && separator.Length == 1)
+                        {
+                            return text.Split(separator[0]);
+                        }
+                    }
+                    else if (string.Equals(_methodMethodName, "PadLeft", StringComparison.OrdinalIgnoreCase))
+                    {
+                        int totalWidth;
+                        string paddingChar;
+                        if (TryGetArg(args, out totalWidth))
+                        {
+                            return text.PadLeft(totalWidth);
+                        }
+                        else if (TryGetArgs(args, out totalWidth, out paddingChar) && paddingChar.Length == 1)
+                        {
+                            return text.PadLeft(totalWidth, paddingChar[0]);
+                        }
+                    }
+                    else if (string.Equals(_methodMethodName, "TrimEnd", StringComparison.OrdinalIgnoreCase))
+                    {
+                        string trimChars;
+                        if (TryGetArg(args, out trimChars) && trimChars.Length > 0)
+                        {
+                            return text.TrimEnd(trimChars.ToCharArray());
```

Comment:
What if the call is `TrimEnd('a', 'b')`? Then this would only preserve the first argument, as a single element char array.

Reply:
Can you please add a test for a multi arg TrimEnd?

---
In reply to: [159286703](https://github.com/Microsoft/msbuild/pull/2833#discussion_r159286703) [](ancestors = 159286703)

Reply:
I have one:
`TestPropertyFunction("$(prop.TrimEnd('.0123456789'))", "prop", "net461", "net");`

This is driven by real usage I encountered in the SDK:
https://github.com/dotnet/sdk/issues/1833

I explicitly want to avoid handling all the cases, so for multi-arg TrimEnd I just want to fall back to the standard binder. I only want to support enough to remove the pain from common cases (Roslyn should evaluate without throwing).

But I could add a test for this scenario, it's just it would test the old behavior. I'm thinking of refactoring a lot of these unit-tests in a separate PR.

Reply:
My main concern with this is having the shortcircuiting method interpret things differently than the legacy code, hence my suggestion to add a unit test for multiple args to ensure it still gets interpreted like before :)

So apparently single quoted strings are valid in MSBuild, and it converts them to char arrays when interpreting the property functions in C# :) . This is most likely a consequence of MSBuild treating single quotes and double quotes equally, and so it cannot discern between char arrays and string arrays. Wonder what happens if a type has overloads of `params string[]` and `params char[]`, probably char[] would win all the time.

Reply:
Added a test and verified that it throws:

```
"C:\Temp\1.proj" (default target) (1) ->
(Build target) ->
  C:\Temp\1.proj(6,18): error MSB4184: The expression ""stringab".TrimEnd(a, b)" cannot be evaluated. Method 'System.String.TrimEnd' not found.
```